File: kit/matcher/find_nested_matches.go
package matcher

import (
	"maps"
	"slices"
	"strings"
)

type FindNestedMatchesResults struct {
	Params      Params
	SplatValues []string
	Matches     []*Match
}

func (m *Matcher) FindNestedMatches(realPath string) (*FindNestedMatchesResults, bool) {
	realPath = StripTrailingSlash(realPath)

	realSegments := ParseSegments(realPath)
	matches := make(matchesMap)

	emptyRR, hasEmptyRR := m.staticPatterns[""]

	if hasEmptyRR {
		matches[emptyRR.normalizedPattern] = &Match{RegisteredPattern: emptyRR}
	}

	realSegmentsLen := len(realSegments)

	if realPath == "" {
		if rr, ok := m.staticPatterns["/"]; ok {
			matches[rr.normalizedPattern] = &Match{RegisteredPattern: rr}
		} else {
			if rr, ok := m.dynamicPatterns["/*"]; ok {
				matches["/*"] = &Match{
					RegisteredPattern: rr,
					splatValues:       []string{},
				}
			}
		}
		return flattenAndSortMatches(matches, realPath, realSegmentsLen)
	}

	var pb strings.Builder
	pb.Grow(len(realPath) + 1)
	var foundFullStatic bool
	for i := range realSegments {
		pb.WriteString("/")
		pb.WriteString(realSegments[i])
		if rr, ok := m.staticPatterns[pb.String()]; ok {
			matches[rr.normalizedPattern] = &Match{RegisteredPattern: rr}
			if i == realSegmentsLen-1 {
				foundFullStatic = true
			}
		}
		if i == realSegmentsLen-1 {
			pb.WriteString("/")
			if rr, ok := m.staticPatterns[pb.String()]; ok {
				matches[rr.normalizedPattern] = &Match{RegisteredPattern: rr}
			}
		}
	}

	if !foundFullStatic {
		// For the catch-all pattern (e.g., "/*"), handle it specially
		if rr, ok := m.dynamicPatterns["/*"]; ok {
			matches["/*"] = &Match{
				RegisteredPattern: rr,
				splatValues:       realSegments,
			}
		}

		// DFS for the rest of the matches
		params := make(Params)
		m.dfsNestedMatches(m.rootNode, realSegments, 0, params, matches)
	}

	// if there are multiple matches and a catch-all, remove the catch-all
	// UNLESS the sole other match is an empty str pattern
	if _, ok := matches["/*"]; ok {
		if hasEmptyRR {
			if len(matches) > 2 {
				delete(matches, "/*")
			}
		} else if len(matches) > 1 {
			delete(matches, "/*")
		}
	}

	if len(matches) < 2 {
		return flattenAndSortMatches(matches, realPath, realSegmentsLen)
	}

	var longestSegmentLen int
	longestSegmentMatches := make(matchesMap)
	for _, match := range matches {
		if len(match.normalizedSegments) > longestSegmentLen {
			longestSegmentLen = len(match.normalizedSegments)
		}
	}
	for _, match := range matches {
		if len(match.normalizedSegments) == longestSegmentLen {
			longestSegmentMatches[match.lastSegType] = match
		}
	}

	// if there is any splat or index with a segment length shorter than longest segment length, remove it
	for pattern, match := range matches {
		if len(match.normalizedSegments) < longestSegmentLen {
			if match.lastSegIsNonRootSplat || match.lastSegIsIndex {
				delete(matches, pattern)
			}
		}
	}

	if len(matches) < 2 {
		return flattenAndSortMatches(matches, realPath, realSegmentsLen)
	}

	// if the longest segment length items are (1) dynamic, (2) splat, or (3) index, remove them as follows:
	// - if the realSegmentLen equals the longest segment length, prioritize dynamic, then splat, and always remove index
	// - if the realSegmentLen is greater than the longest segment length, prioritize splat, and always remove dynamic and index
	if len(longestSegmentMatches) > 1 {
		if match, indexExists := longestSegmentMatches[segTypes.index]; indexExists {
			delete(matches, match.normalizedPattern)
		}

		_, dynamicExists := longestSegmentMatches[segTypes.dynamic]
		_, splatExists := longestSegmentMatches[segTypes.splat]

		if realSegmentsLen == longestSegmentLen && dynamicExists && splatExists {
			delete(matches, longestSegmentMatches[segTypes.splat].normalizedPattern)
		}
		if realSegmentsLen > longestSegmentLen && splatExists && dynamicExists {
			delete(matches, longestSegmentMatches[segTypes.dynamic].normalizedPattern)
		}
	}

	return flattenAndSortMatches(matches, realPath, realSegmentsLen)
}

func (m *Matcher) dfsNestedMatches(
	node *segmentNode,
	segments []string,
	depth int,
	params Params,
	matches matchesMap,
) {
	if len(node.pattern) > 0 {
		if rp := m.dynamicPatterns[node.pattern]; rp != nil {
			// Don't process the ultimate catch-all here
			if node.pattern != "/*" {
				// Copy params
				paramsCopy := make(Params, len(params))
				maps.Copy(paramsCopy, params)

				var splatValues []string
				if node.nodeType == nodeSplat && depth < len(segments) {
					// For splat nodes, collect all remaining segments
					splatValues = make([]string, len(segments)-depth)
					copy(splatValues, segments[depth:])
				}

				match := &Match{
					RegisteredPattern: rp,
					params:            paramsCopy,
					splatValues:       splatValues,
				}
				matches[node.pattern] = match

				// Check for index segment if we're at the exact depth
				if depth == len(segments) {
					var sb strings.Builder
					sb.Grow(len(node.pattern) + 1)
					sb.WriteString(node.pattern)
					sb.WriteByte('/')
					indexPattern := sb.String()
					if rp, ok := m.dynamicPatterns[indexPattern]; ok {
						matches[indexPattern] = &Match{
							RegisteredPattern: rp,
							params:            paramsCopy,
						}
					}
				}
			}
		}
	}

	// If we've consumed all segments, stop
	if depth >= len(segments) {
		return
	}

	seg := segments[depth]

	// Try static children
	if node.children != nil {
		if child, ok := node.children[seg]; ok {
			m.dfsNestedMatches(child, segments, depth+1, params, matches)
		}
	}

	// Try dynamic/splat children
	for _, child := range node.dynChildren {
		switch child.nodeType {
		case nodeDynamic:
			// Backtracking pattern for dynamic
			oldVal, hadVal := params[child.paramName]
			params[child.paramName] = seg

			m.dfsNestedMatches(child, segments, depth+1, params, matches)

			if hadVal {
				params[child.paramName] = oldVal
			} else {
				delete(params, child.paramName)
			}

		case nodeSplat:
			// For splat nodes, we collect remaining segments and don't increment depth
			m.dfsNestedMatches(child, segments, depth, params, matches)
		}
	}
}

func flattenAndSortMatches(
	matches matchesMap,
	realPath string,
	realSegmentLen int,
) (*FindNestedMatchesResults, bool) {
	var results []*Match
	for _, match := range matches {
		results = append(results, match)
	}

	slices.SortStableFunc(results, func(i, j *Match) int {
		// if any match is an index, it should be last
		if i.lastSegIsIndex {
			return 1
		}
		if j.lastSegIsIndex {
			return -1
		}

		// else sort by segment length
		lenDiff := len(i.normalizedSegments) - len(j.normalizedSegments)
		if lenDiff != 0 {
			return lenDiff
		}

		// Tiebreaker for determinism when segment lengths are equal
		return strings.Compare(i.normalizedPattern, j.normalizedPattern)
	})

	if len(results) == 0 {
		return nil, false
	}

	// if not slash route and solely matched "", then invalid
	isNotSlashRoute := realPath != "" && realPath != "/"
	if isNotSlashRoute && len(results) == 1 && results[0].normalizedPattern == "" {
		return nil, false
	}

	lastMatch := results[len(results)-1]

	// Check if the last match can consume all segments.
	if !lastMatch.lastSegIsNonRootSplat && lastMatch.normalizedPattern != "/*" {
		// For non-splat patterns, check if pattern depth matches
		// real segment count.
		// Dynamic patterns should have filled params if they matched.
		patternSegmentsLen := len(lastMatch.normalizedSegments)

		// If pattern has fewer segments than the real path and it's
		// not a splat, it's a partial match.
		if patternSegmentsLen < realSegmentLen {
			return nil, false
		}

		// If it's a dynamic pattern at the right depth, it should
		// have params.
		if patternSegmentsLen == realSegmentLen &&
			lastMatch.numberOfDynamicParamSegs > 0 &&
			len(lastMatch.params) == 0 {
			return nil, false
		}
	}

	return &FindNestedMatchesResults{
		Params:      lastMatch.params,
		SplatValues: lastMatch.splatValues,
		Matches:     results,
	}, true
}

File: typescript/kit/matcher/find_nested_matches.ts
import { parseSegments } from "./parse_segments.ts";
import type {
	Params,
	PatternRegistry,
	RegisteredPattern,
	SegmentNode,
} from "./register.ts";
import { NODE_DYNAMIC, NODE_SPLAT, SEG_TYPES } from "./register.ts";

export type Match = {
	registeredPattern: RegisteredPattern;
	params: Params;
	splatValues: string[];
};

type FindNestedMatchesResult = {
	params: Params;
	splatValues: string[];
	matches: Match[];
};

function stripTrailingSlash(pattern: string): string {
	return pattern.length > 0 && pattern[pattern.length - 1] === "/"
		? pattern.substring(0, pattern.length - 1)
		: pattern;
}

function dfsNestedMatches(
	registry: PatternRegistry,
	node: SegmentNode,
	segments: string[],
	depth: number,
	params: Params,
	matches: Map<string, Match>,
): void {
	if (node.pattern.length > 0) {
		const rp = registry.dynamicPatterns.get(node.pattern);
		if (rp && node.pattern !== "/*") {
			const paramsCopy = { ...params };
			let splatValues: string[] = [];

			if (node.nodeType === NODE_SPLAT && depth < segments.length) {
				splatValues = segments.slice(depth);
			}

			matches.set(node.pattern, {
				registeredPattern: rp,
				params: paramsCopy,
				splatValues,
			});

			if (depth === segments.length) {
				const indexPattern = node.pattern + "/";
				const rpIndex = registry.dynamicPatterns.get(indexPattern);
				if (rpIndex) {
					matches.set(indexPattern, {
						registeredPattern: rpIndex,
						params: paramsCopy,
						splatValues: [],
					});
				}
			}
		}
	}

	if (depth >= segments.length) return;

	const seg = segments[depth]!;

	if (node.children !== null) {
		const child = node.children.get(seg);
		if (child) {
			dfsNestedMatches(
				registry,
				child,
				segments,
				depth + 1,
				params,
				matches,
			);
		}
	}

	for (const child of node.dynChildren) {
		switch (child.nodeType) {
			case NODE_DYNAMIC: {
				const oldVal = params[child.paramName];
				const hadVal = oldVal !== undefined;
				params[child.paramName] = seg;

				dfsNestedMatches(
					registry,
					child,
					segments,
					depth + 1,
					params,
					matches,
				);

				if (hadVal) {
					params[child.paramName] = oldVal!;
				} else {
					delete params[child.paramName];
				}
				break;
			}
			case NODE_SPLAT:
				dfsNestedMatches(
					registry,
					child,
					segments,
					depth,
					params,
					matches,
				);
				break;
		}
	}
}

function flattenAndSortMatches(
	matches: Map<string, Match>,
	realPath: string,
	realSegmentLen: number,
): FindNestedMatchesResult | null {
	const results: Match[] = Array.from(matches.values());

	results.sort((i, j) => {
		if (i.registeredPattern.lastSegIsIndex) return 1;
		if (j.registeredPattern.lastSegIsIndex) return -1;
		return (
			i.registeredPattern.normalizedSegments.length -
			j.registeredPattern.normalizedSegments.length
		);
	});

	if (results.length === 0) return null;

	const isNotSlashRoute = realPath !== "" && realPath !== "/";
	if (
		isNotSlashRoute &&
		results.length === 1 &&
		results[0]!.registeredPattern.normalizedPattern === ""
	) {
		return null;
	}

	const lastMatch = results[results.length - 1]!;

	if (
		!lastMatch.registeredPattern.lastSegIsNonRootSplat &&
		lastMatch.registeredPattern.normalizedPattern !== "/*"
	) {
		const patternSegmentsLen =
			lastMatch.registeredPattern.normalizedSegments.length;

		if (patternSegmentsLen < realSegmentLen) return null;

		if (
			patternSegmentsLen === realSegmentLen &&
			lastMatch.registeredPattern.numberOfDynamicParamSegs > 0 &&
			Object.keys(lastMatch.params).length === 0
		) {
			return null;
		}
	}

	return {
		params: lastMatch.params,
		splatValues: lastMatch.splatValues,
		matches: results,
	};
}

export function findNestedMatches(
	registry: PatternRegistry,
	realPath: string,
): FindNestedMatchesResult | null {
	realPath = stripTrailingSlash(realPath);

	const realSegments = parseSegments(realPath);
	const matches: Map<string, Match> = new Map();

	// Check for empty pattern
	const emptyRR = registry.staticPatterns.get("");
	if (emptyRR) {
		matches.set(emptyRR.normalizedPattern, {
			registeredPattern: emptyRR,
			params: {},
			splatValues: [],
		});
	}

	const realSegmentsLen = realSegments.length;

	// Handle root path
	if (realPath === "") {
		const rr = registry.staticPatterns.get("/");
		if (rr) {
			matches.set(rr.normalizedPattern, {
				registeredPattern: rr,
				params: {},
				splatValues: [],
			});
		} else {
			const catchAllRR = registry.dynamicPatterns.get("/*");
			if (catchAllRR) {
				matches.set("/*", {
					registeredPattern: catchAllRR,
					params: {},
					splatValues: [],
				});
			}
		}
		return flattenAndSortMatches(matches, realPath, realSegmentsLen);
	}

	// Check static patterns progressively
	let pb = "";
	let foundFullStatic = false;

	for (let i = 0; i < realSegments.length; i++) {
		pb += "/" + realSegments[i];
		const rr = registry.staticPatterns.get(pb);
		if (rr) {
			matches.set(rr.normalizedPattern, {
				registeredPattern: rr,
				params: {},
				splatValues: [],
			});
			if (i === realSegmentsLen - 1) foundFullStatic = true;
		}
		if (i === realSegmentsLen - 1) {
			pb += "/";
			const rrWithSlash = registry.staticPatterns.get(pb);
			if (rrWithSlash) {
				matches.set(rrWithSlash.normalizedPattern, {
					registeredPattern: rrWithSlash,
					params: {},
					splatValues: [],
				});
			}
		}
	}

	// Check dynamic patterns if no full static match
	if (!foundFullStatic) {
		const rr = registry.dynamicPatterns.get("/*");
		if (rr) {
			matches.set("/*", {
				registeredPattern: rr,
				params: {},
				splatValues: realSegments,
			});
		}

		const params: Params = {};
		dfsNestedMatches(
			registry,
			registry.rootNode,
			realSegments,
			0,
			params,
			matches,
		);
	}

	// Clean up catch-all pattern if necessary
	const hasEmptyRR = emptyRR !== undefined;
	if (matches.has("/*")) {
		if (hasEmptyRR) {
			if (matches.size > 2) matches.delete("/*");
		} else if (matches.size > 1) {
			matches.delete("/*");
		}
	}

	if (matches.size < 2) {
		return flattenAndSortMatches(matches, realPath, realSegmentsLen);
	}

	// Find longest segment matches
	let longestSegmentLen = 0;
	const longestSegmentMatches: Map<string, Match> = new Map();

	for (const match of matches.values()) {
		if (
			match.registeredPattern.normalizedSegments.length >
			longestSegmentLen
		) {
			longestSegmentLen =
				match.registeredPattern.normalizedSegments.length;
		}
	}

	for (const match of matches.values()) {
		if (
			match.registeredPattern.normalizedSegments.length ===
			longestSegmentLen
		) {
			longestSegmentMatches.set(
				match.registeredPattern.lastSegType,
				match,
			);
		}
	}

	// Remove shorter matches
	for (const [pattern, match] of matches) {
		if (
			match.registeredPattern.normalizedSegments.length <
			longestSegmentLen
		) {
			if (
				match.registeredPattern.lastSegIsNonRootSplat ||
				match.registeredPattern.lastSegIsIndex
			) {
				matches.delete(pattern);
			}
		}
	}

	if (matches.size < 2) {
		return flattenAndSortMatches(matches, realPath, realSegmentsLen);
	}

	// Handle conflicts between longest matches
	if (longestSegmentMatches.size > 1) {
		const indexMatch = longestSegmentMatches.get(SEG_TYPES.index);
		if (indexMatch) {
			matches.delete(indexMatch.registeredPattern.normalizedPattern);
		}

		const dynamicExists = longestSegmentMatches.has(SEG_TYPES.dynamic);
		const splatExists = longestSegmentMatches.has(SEG_TYPES.splat);

		if (
			realSegmentsLen === longestSegmentLen &&
			dynamicExists &&
			splatExists
		) {
			const splatMatch = longestSegmentMatches.get(SEG_TYPES.splat)!;
			matches.delete(splatMatch.registeredPattern.normalizedPattern);
		}

		if (
			realSegmentsLen > longestSegmentLen &&
			splatExists &&
			dynamicExists
		) {
			const dynamicMatch = longestSegmentMatches.get(SEG_TYPES.dynamic)!;
			matches.delete(dynamicMatch.registeredPattern.normalizedPattern);
		}
	}

	return flattenAndSortMatches(matches, realPath, realSegmentsLen);
}
